File: abstract-factory/go/src/ShopFactory.go
package src

import (
  "microwind/src/shop"
  "microwind/src/vehicle"
)

type ShopFactory struct {
}

func (s *ShopFactory) GetShop(shopType string, name string) shop.Shop {
  switch shopType {
  case "AgencyShop":
    return &shop.AgencyShop{
      Name: name,
    }
  case "DirectSaleShop":
    return &shop.DirectSaleShop{
      Name: name,
    }
  case "SupermarketShop":
    return &shop.SupermarketShop{
      Name: name,
    }
  default:
    return &shop.AgencyShop{
      Name: name,
    }
  }
}

func (s *ShopFactory) GetVehicle(vehicleType string, name string) vehicle.Vehicle {
  // ignored, this function for Vehicle
  return nil
}

File: abstract-factory/java/src/ShopFactory.java
package src;
import src.vehicle.*;
import src.shop.*;
/*
 商店创建类，继承自抽象工厂
*/
public class ShopFactory extends AbstractFactory {
  @Override
  public Shop getShop(String name) {
    if (name.equals("AgencyShop")) {
      return new AgencyShop();
    } else if (name.equals("DirectSaleShop")) {
      return new DirectSaleShop();
    } else if (name.equals("SupermarketShop")) {
      return new SupermarketShop();
    } else {
      return null;
    }
  }

  @Override
  public AbstractVehicle getVehicle(VehicleType type) {
    return null;
  }
}
